### ISSUE TITLE
Relative classpath resource lookup fails in native image

### ISSUE THREAD

**Issue by sbrannen** [oracle/graal#2250]
# Overview

If the path supplied to `getClass().getResource()` is a relative path, lookup for the resource fails in a compiled native image, even when using the configuration generated by the agent.

For example, given the `RelativeClasspathResource` test case below, the output when using a standard JDK results in the following.

```
SUCCESS: found and read classpath resource: subdir/example1.txt
SUCCESS: found and read classpath resource: example2.txt
SUCCESS: found and read classpath resource: subdir/../example2.txt
```

Whereas, if you compile the test case into a native image, the output is the following.

```
FAILURE: could not find classpath resource: subdir/example1.txt
FAILURE: could not find classpath resource: example2.txt
FAILURE: could not find classpath resource: subdir/../example2.txt
```

If you execute the test case with the GraalVM native image agent, it generates the following in `resource-config.json`:

```json
{
  "resources":[
    {"pattern":"\\Qexample2.txt\\E"}, 
    {"pattern":"\\Qsubdir/../example2.txt\\E"}, 
    {"pattern":"\\Qsubdir/example1.txt\\E"}
  ],
  "bundles":[]
}
```

If you then build the native image with the configuration generated by the agent, the output is the following.

```
SUCCESS: found and read classpath resource: subdir/example1.txt
SUCCESS: found and read classpath resource: example2.txt
FAILURE: could not find classpath resource: subdir/../example2.txt
```

The above **FAILURE** points out the bug within a native image.

The first two resources are now found; however, the third (relative path) resource is still not found. Note that the agent generates an entry in `resource-config.json` for `{"pattern":"\\Qsubdir/../example2.txt\\E"}`. Thus, it appears that the relative path (i.e., `subdir/../`) is not properly supported within a native image.

See https://github.com/sbrannen/graalvm-playground/tree/master/relative_classpath_resource for full details.

```java
public class RelativeClasspathResource {

	public static void main(String[] args) throws Exception {
		new Tester().test();
	}


	static class Tester {

		void test() throws Exception {
			classpathResource("subdir/example1.txt", "1");
			classpathResource("example2.txt", "2");
			classpathResource("subdir/../example2.txt", "2");
		}

		private void classpathResource(String name, String expectedContents) throws Exception {
			URL resource = getClass().getResource(name);
			if (resource == null) {
				System.err.println("FAILURE: could not find classpath resource: " + name);
				return;
			}
			String contents = getContentsAsString(getClass().getResourceAsStream(name));
			if (contents == null) {
				System.err.println("FAILURE: could not read contents of classpath resource: " + name);
				return;
			}
			if (!contents.equals(expectedContents)) {
				System.err.format("FAILURE: classpath resource '%s' contains '%s' instead of '%s'%n", name, contents,
					expectedContents);
			}
			System.err.println("SUCCESS: found and read classpath resource: " + name);
		}
	}


	private static String getContentsAsString(InputStream in) throws Exception {
		try (BufferedReader reader = new BufferedReader(new InputStreamReader(in))) {
			return reader.readLine();
		}
	}

}
```

# Environment

- GraalVM version: CE 20.0
- JDK major version: 8
- OS: macOS 10.14.6
- Architecture: i386 / x86_64

**Comment by peter-hofer** [comment]
Thank you for your report @sbrannen ! Indeed resource lookups can currently only be done with the exact path in the configuration.

**Comment by sbrannen** [comment]
> Indeed resource lookups can currently only be done with the exact path in the configuration.

The exact path was already in the configuration: `{"pattern":"\\Qexample2.txt\\E"}`.

The issue back then was that the invocation of `getClass().getResourceAsStream("subdir/../example2.txt")` failed to find the resource.

Good news: that is no longer an issue!

I just tested this again with `OpenJDK Runtime Environment GraalVM CE 22.2.0 (build 17.0.4+8-jvmci-22.2-b06)`, and the relative resource is now found via `Class#getResourceAsStream`. So that obviously got fixed in the meantime. 👍 

However, the tracing agent still produces incorrect resource patterns.

```json
{
  "resources":[
    {"pattern":"\\Qexample2.txt\\E"}, 
    {"pattern":"\\Qsubdir/../example2.txt\\E"}, 
    {"pattern":"\\Qsubdir/example1.txt\\E"}
  ],
  "bundles":[]
}
```

The second pattern `"pattern":"\\Qsubdir/../example2.txt\\E"` should not be generated in `resource-config.json`.

In light of the above, I'll close this issue and open a new one regarding the agent.
